### PR TITLE
GH-32276: [C++][FlightRPC] Align buffers from Flight

### LINKED ISSUE #32276: [C++][FlightRPC] Flight generates misaligned buffers
Protobuf's wire format design + our zero-copy serializer/deserializer mean that buffers can end up misaligned. On some Arrow versions, this can cause segfaults in kernels assuming alignment (and generally violates expectations). 

We should:
- Possibly include buffer alignment in array validation
- See if we can adjust the serializer to somehow pad things properly
- See if we can do anything about this in the deserializer
  
  Example:
  ```python
  
  import pyarrow as pa
  import pyarrow.flight as flight
  
  class TestServer(flight.FlightServerBase):
      def do_get(self, context, ticket):
          schema = pa.schema(
              [
                  ("index", pa.int64()),
                  ("int8", pa.float64()),
                  ("int16", pa.float64()),
                  ("int32", pa.float64()),
              ]
          )
          return flight.RecordBatchStream(pa.table([
              [0, 1, 2, 3],
              [0, 1, None, 3],
              [0, 1, 2, None],
              [0, None, 2, 3],
          ], schema=schema))
  
  
  with TestServer() as server:
      client = flight.connect(f"grpc://localhost:{server.port}")
      table = client.do_get(flight.Ticket(b"")).read_all()
      for col in table:
          print(col.type)
          for chunk in col.chunks:
              for buf in chunk.buffers():
                  if not buf: continue
                  print("buffer is 8-byte aligned?", buf.address % 8)
              chunk.cast(pa.float32())
  ```
  
  On Arrow 8
  ```
  
  int64
  buffer is 8-byte aligned? 1
  double
  buffer is 8-byte aligned? 1
  buffer is 8-byte aligned? 1
  double
  buffer is 8-byte aligned? 1
  buffer is 8-byte aligned? 1
  double
  buffer is 8-byte aligned? 1
  buffer is 8-byte aligned? 1
  ```
  On Arrow 7
  ```
  
  int64
  buffer is 8-byte aligned? 4
  double
  buffer is 8-byte aligned? 4
  buffer is 8-byte aligned? 4
  fish: Job 1, 'python ../test.py' terminated by signal SIGSEGV (Address boundary error)
  ```

**Reporter**: [David Li](https://issues.apache.org/jira/browse/ARROW-16958) / @lidavidm

<sub>**Note**: *This issue was originally created as [ARROW-16958](https://issues.apache.org/jira/browse/ARROW-16958). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by lidavidm**
### Rationale for this change

Ensure buffers from Flight are at least 8-byte aligned to meet general expectations of downstream users.

### What changes are included in this PR?

Manually align buffers after deserialization.

### Are these changes tested?

Yes

### Are there any user-facing changes?

Flight data will be aligned, at the cost of an additional copy.
* Closes: #32276

### CODE REVIEW

**Comment by lidavidm** [comment]
I inlined some stuff from @westonpace's PR but I'd rather wait and be able to use those utilities directly. That said it demonstrates the problem and the solution. 

Given more time we might be able to avoid this, but I'd need to dig deeper into Protobuf & do some benchmarking. (In particular: instead of going gRPC slices -> CodedInputStream -> Arrow buffers, I'd rather directly go from gRPC slices -> Arrow.)

File: cpp/src/arrow/flight/transport/grpc/serialization_internal.cc [hunk 1]
```diff
@@ -380,6 +382,14 @@ ::grpc::Status FlightDataDeserialize(ByteBuffer* buffer,
           return ::grpc::Status(::grpc::StatusCode::INTERNAL,
                                 "Unable to read FlightData body");
         }
+        // XXX: due to where we sit, we can't use a custom allocator
+        // XXX: any error here will likely crash or hang gRPC!
+        auto status =
+            util::EnsureAlignment(std::move(out->body), 64, default_memory_pool())
```
Comment (pitrou) [inline cpp/src/arrow/flight/transport/grpc/serialization_internal.cc:388]: Isn't 64 a bit heavy-handed? Basically, and assuming the distribution of gRPC slice alignments is uniform, we will reallocate almost all incoming buffers?

File: cpp/src/arrow/flight/transport/grpc/serialization_internal.cc [hunk 2]
```diff
@@ -380,6 +382,14 @@ ::grpc::Status FlightDataDeserialize(ByteBuffer* buffer,
           return ::grpc::Status(::grpc::StatusCode::INTERNAL,
                                 "Unable to read FlightData body");
         }
+        // XXX: due to where we sit, we can't use a custom allocator
+        // XXX: any error here will likely crash or hang gRPC!
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/arrow/flight/transport/grpc/serialization_internal.cc:386]: This is true of any error returned here. Last I checked, gRPC doesn't actually handle errors despite this being a fallible API, instead choosing to crash.

**Comment by lidavidm** [comment]
Given the nature of the gRPC API there's not a great way to customize it at runtime. If someone complains about this in the future we can point them here then. Possibly we can make it a compile-time toggle, though that's effectively useless to most users.

@westonpace @rtpsw were the most recent to mention this. If there's not a real desire for this then I'll close the PR.

**Comment by pitrou** [comment]
This could be settable using an env var. But I don't think we should realign buffers by default. This can decrease performance and increase memory fragmentation for unsuspecting users.

**Comment by westonpace** [comment]
Now that we have a workaround enabled in Acero I don't have any urgent need for this.  I'm fine with waiting for it to become a problem (and it may never do so).

**Comment by orlp** [comment]
@pitrou 

> But I don't think we should realign buffers by default. 

The [arrow standard requires it](https://arrow.apache.org/docs/format/Columnar.html#buffer-alignment-and-padding).

> Implementations are recommended to allocate memory on aligned addresses (multiple of 8- or 64-bytes) and pad (overallocate) to a length that is a multiple of 8 or 64 bytes. When serializing Arrow data for interprocess communication, these alignment and padding requirements are **enforced**.

File: cpp/src/arrow/flight/transport/grpc/serialization_internal.cc (hunk 1 above)
Comment (EnricoMi) [reply cpp/src/arrow/flight/transport/grpc/serialization_internal.cc:388]: This should be configurable as not all code using the C++ flight client is sensitive to alignment.

**Pull request closed without merging**